### ISSUE TITLE
Feature Wishlist

### ISSUE THREAD

**Issue by NAThompson** [boostorg/math#303]
- Modified Clenshaw recurrence for all special functions governed by 3-term recurrence relations.
- Higher dimensional interpolation (Akima has a natural generalization to higher dimensions, but there's some work on getting the API correct.)
- Prolate spheroidal wavefunctions
- Higher dimensional quadrature; sparse grids for medium dimensionality, extensions of Gaussian quadrature in lower dimensions.
- Wavelet transforms, shearlets
- Bivariate interpolation and quadrature at the [Padua points](https://www.math.unipd.it/~marcov/pdf/poster5ecm.pdf)
- [Von Mises distribution](https://github.com/boostorg/math/issues/296)
- [Gauss-Hermite quadrature](https://github.com/boostorg/math/issues/1453)
- [Generalized Laguerre polynomials](https://github.com/boostorg/math/issues/1128)
- [Interpolation with different units](https://github.com/boostorg/math/issues/314)

What else is a natural fit for the library, and is useful?

**Comment by ckormanyos** [comment]
> What else is a natural fit for the library, and is useful?

* Make further progress on zeta, phi and friends (Lerch phi transcendent, polylog, and more).
* Extend constants to higher precision such as 10,000 decimal digits and get them tested.
* Evaluate recent progress in the literature on Bernoulli number generation (can we save memory and/or time compared with extracting Bn from tangent numbers?).

**Comment by ckormanyos** [comment]
> What else is a natural fit for the library, and is useful?

* Legendre functions and associated Legendre functions of arbitrary (non-integer real) order and degree.
* More special functions of complex argument, particularly on the unit disk. Gamma should be a reasonable place to start. Perhaps complex orthogonal polynomials if these are calculated from straightforward recursion.
* FFT

**Comment by pabristow** [comment]
Would it be useful to post a introduction message and link to this 

[https://github.com/boostorg/math/issues/303 
](url)

on the Boost.users and developers lists to try to get input from a wider audience?

**Comment by cosurgi** [comment]
* Definitely FFT, best if you used techniques employed by fftw. I tried using libfftw with float128 and although library claims to work, the precision (at first try) seemed to be reduced to double.

* I found a bug in tgamma function for float128, and had to use the following workaround:
https://gitlab.com/yade-dev/trunk/blob/highPrecisionReal/lib/high-precision/MathFunctions.hpp#L239 (maybe you already had this fixed in latest boost version though)

I have just recently integratated boost multiprecision with a fairly large numerical computation software yade: https://gitlab.com/yade-dev/trunk/merge_requests/362

**Comment by NAThompson** [comment]
@cosurgi : It seems like most of your requirements are actually from boost.multiprecision, rather than boost.math. Do you mind reposting over there?

**Comment by cosurgi** [comment]
Thank you, should I open a new issue in https://github.com/boostorg/multiprecision/issues ?

**Comment by NAThompson** [comment]
Yeah, the tgamma should probably be a separate issue against boost.math. quad-double is for multiprecision.

**Comment by ckormanyos** [comment]
> Thank you, should I open a new issue in https://github.com/boostorg/multiprecision/issues ?

Sure. That seems like a good idea. If there is an issue lie that, I will add to it.

**Comment by ckormanyos** [comment]
> How feasible it is to integrate quad-double precision...

This is a good question. I would combine that with also software support for float128_t for those platforms lacking quadmath. In fact, I would see full cross-plattform availability of a float128_t (be it hardware backend or stware pure) as a prerequisite before going for a float256_t.

Then there is the topc of design questions.
* Exrtract float128_t/256_t from the same template?
* Support other digit splits, etc.

Lots of questions.

Kind regards, Chris.

**Comment by HDembinski** [comment]
To second @NAThompson proposal to add ND Akima splines, matlab has a nice modified version called "Makima" which works for pdfs and cdfs (this was brought to my attention by Nick, too)
https://blogs.mathworks.com/cleve/2019/04/29/makima-piecewise-cubic-interpolation/

Matlab also have a nice spline algorithm called pchip, which avoids undulations of the splines outside of the range of the data points, which often desired in practice.

**Comment by pabristow** [comment]
It would be nice if float128_t could magically use float128 (assembler software) when available (GCC only at present - but should be possible for Clang too but I haven't been able to get it up the developers priority pile for any work on making it happen).

**Comment by cosurgi** [comment]
Thank you very much everyone. I have submitted separate issues for

* [bug in tgamma float128 #307](https://github.com/boostorg/math/issues/307)
* and for [quad-double](https://github.com/boostorg/multiprecision/issues/184)

Regarding the FFT, here's how I tried `fftwq` for float128:

source file `test_fftwq.cpp`:

```cpp
#include <complex>
#include <limits>
#include <iostream>
#include <boost/cstdfloat.hpp>
#include <boost/math/cstdfloat/cstdfloat_types.hpp>
#include <boost/multiprecision/float128.hpp>
#include <cmath>
#include <fftw3.h>

int main() {
	using Real = boost::float128_t;
	using Complex = std::complex<Real>;

	std::vector<Complex> in {{1,-1}, { 0 , 0 }, { 1  , 0.5} , { 0  , 0.5 } , {0.3Q, 1.1Q} };
	std::vector<Complex> out{{0, 0}, { 0 , 0 }, { 0  , 0  } , { 0  , 0   } , { 0  ,  0  } };
	size_t N=in.size();
	fftwq_plan p;
	p = fftwq_plan_dft_1d(N , (Real (*)[2])(in.data()) , (Real (*)[2])(out.data()) , FFTW_FORWARD, FFTW_ESTIMATE);
	fftwq_execute(p);
	fftwq_destroy_plan(p);

	int digs = std::numeric_limits<Real>::digits10;
	std::cout << "precision digits=" << digs << "\n";
	std::cout << std::setprecision(digs+1);
	for(size_t i = 0 ; i < N ; ++i) { std::cout << in [i] << "\n"; }
	std::cout << "---\n";
	for(size_t i = 0 ; i < N ; ++i) { std::cout << (out[i]/std::sqrt(Real(N))) << "\n"; }
	std::cout << "---\n";
	std::cout << std::sqrt(Real(3.Q)) << "\n";

}
```

the `makefile`:
```makefile
all:
	g++ -O1 -g test_fftwq.cpp -o test_fftwq -std=gnu++17 \
	-Wall -Werror=narrowing -lquadmath -lfftw3q -lfftw3l -lfftw3
run:
	test_fftwq
```

Program output:

```
precision digits=33
(1,-1)
(0,0)
(1,0.5)
(0,0.5)
(0.3,1.1)
---
(1.028591269649903260348219887616387,0.4919349550499537332100182071208808)
(-0.3409887676311503536846088281059338,-0.7922686679441966705904014368275257)
(0.1877170652969481257011204486123902,-0.2028156621435458504101773577043522)
(0.7660212886279950923293564419450544,-1.211185804131325864399418580221726)
(0.5947271215560935717150857186633783,-0.5217327983306750442191945010985525)
---
1.732050807568877293527446341505872
```

mathematica notebook for verification:

```
$MinPrecision = 35
$MaxPrecision = 10*$MinPrecision
$MaxExtraPrecision = 20*$MinPrecision

DecimalForm[Fourier[
  SetPrecision[
   {1 - 1 I,
    0 + 0 I,
    1 + 0.5 I,
    0 + 0.5 I,
    0.3 + 1.1 I
    }, $MinPrecision]], DefaultPrintPrecision -> $MinPrecision]

N[Sqrt[3], $MinPrecision]
```

mathematica output:

```
{1.0285912696499032553831515811218411+0.49193495504995377293056465907724873 I,
0.59472712155609360795727976766156939-0.52173279833067502722281005484233306 I,
0.76602128862799511969333249758965442-1.2111858041313258936156217585316901 I,
0.18771706529694810637079366934084586-0.20281566214354588546316839037883404 I,
-0.34098876763115039299538384698263451-0.79226866794419666303813812405566781 I}

1.7320508075688772935274463415058724
```

As you can see the incorrect results start after 18th numerical place, which suggests that fftwq somewhere along the way gets cut to long double. The different order of numbers is something I didn't investigate, but surely the ordering of results in table is explained in respective documentations. It is quite common to shift by one or rotate the numbers in the tables. It all depends on what is most efficient for respective FFT algorithms.

**Comment by NAThompson** [comment]
@cosurgi : Have you submitted this bug against FFTW?

**Comment by cosurgi** [comment]
@NAThompson yeah, thanks, I just did that: https://github.com/FFTW/fftw3/issues/188

**Comment by NAThompson** [comment]
Isn't there some UB here? `std::complex` is only defined on `float`, `double`, and `long double`. I think you have to use the custom fftw 128 bit complex type.

**Comment by cosurgi** [comment]
I assumed that C++ standard requirement that complex type can be reinterpret_casted to an array[2] of real types was sufficient here. And also I wanted to stay within `boost::float128_t` realm. I have just changed this example to use `fftwq_complex`, however that's not what I would like to have in the end. Using `boost::multiprecision` everywhere without exception is my goal.

Here's a modified code (`makefile` remains the same):

```cpp
#include <complex>
#include <limits>
#include <array>
#include <iostream>
#include <boost/cstdfloat.hpp>
#include <boost/math/cstdfloat/cstdfloat_types.hpp>
#include <boost/multiprecision/float128.hpp>
#include <cmath>
#include <fftw3.h>

int main() {
	using Real = __float128;//boost::float128_t;		// fftwq_
	using Complex = fftwq_complex;//std::complex<Real>;

	std::array<Complex,5> in;
	in[0][0]= 1.0Q;
	in[0][1]=-1.0Q;
	in[1][0]= 0.0Q;
	in[1][1]= 0.0Q;
	in[2][0]= 1.0Q;
	in[2][1]= 0.5Q;
	in[3][0]= 0.0Q;
	in[3][1]= 0.5Q;
	in[4][0]= 0.3Q;
	in[4][1]= 1.1Q;
	std::array<Complex,5> out;
	size_t N=in.size();
	fftwq_plan p;
	p = fftwq_plan_dft_1d(N , (Real (*)[2])(in.data()) , (Real (*)[2])(out.data()) , FFTW_FORWARD, FFTW_ESTIMATE);
	fftwq_execute(p);
	fftwq_destroy_plan(p);

	int digs = std::numeric_limits<Real>::digits10;
	std::cout << "precision digits=" << digs << "\n";
	std::cout << std::setprecision(digs+1);
	for(size_t i = 0 ; i < N ; ++i) { std::cout << in [i][0] << " , " << in[i][1] << "\n"; }
	std::cout << "---\n";
	for(size_t i = 0 ; i < N ; ++i) { std::cout << (out[i][0]/sqrt(__float128(N))) << " , " << (out[i][1]/sqrt(__float128(N))) << "\n"; }
	std::cout << "---\n";
	std::cout << std::sqrt(Real(3.Q)) << "\n";

}
```

I obtain the same incorrect output:

```
precision digits=33
1 , -1
0 , 0
1 , 0.5
0 , 0.5
0.3 , 1.1
---
1.028591269649903260348219887616387 , 0.4919349550499537332100182071208808
-0.3409887676311503536846088281059338 , -0.7922686679441966705904014368275257
0.1877170652969481257011204486123902 , -0.2028156621435458504101773577043522
0.7660212886279950923293564419450544 , -1.211185804131325864399418580221726
0.5947271215560935717150857186633783 , -0.5217327983306750442191945010985525
---
1.732050807568877293527446341505872
```

**Comment by stevengj** [comment]
As explained in FFTW/fftw3#188, FFTW's output is correct (it is giving full quad precision as desired) — you are using Mathematica incorrectly.

**Comment by cosurgi** [comment]
I'm sorry. The issue is solved in https://github.com/FFTW/fftw3/issues/188 the problem was due to me not having enough skills in using mathematica. I will remember to never write decimal representation in mathematica. Usually I avoid that, but I thought that explicitly specifying `$MinPrecision` is enough to be able to write decimal floating point in mathematica.

**Comment by cosurgi** [comment]
However still it would be great to have FFT in boost math :)

**Comment by ckormanyos** [comment]
> write decimal representation in mathematica

You can also use inputs like 1.1`33 for 11/10 with 33 decimal digits of precision.

> However still it would be great to have FFT in boost math

Yes, agreed, and a tough challenge...
Thanks for the good suggestions. Kind regards, Chris

**Comment by NAThompson** [comment]
Friends, if you don't mind, I'm going to clean up this thread since it kinda got off topic and many of the off-topic issues have been resolved.

**Comment by octopus-prime** [comment]
- fixed sized polynomials without allocation
- corresponding polynomial solvers, e.g. analytic, bairstow, companion matrix

**Comment by NAThompson** [comment]
@octopus-prime : Yup. Also that reminds me: Jenkins-Traub.

**Comment by cosurgi** [comment]
Did you notice recent discussion about [FFT in mailing list](https://lists.boost.org/Archives/boost/2020/02/248146.php)? I think it would be great if math supported FFT for all multiprecision types, especially [quad-double, which is currently in the works](https://github.com/boostorg/multiprecision/issues/184). And please note that it's not possible to use libfftw as a backend for quad-double.

**Comment by sebweb3r** [comment]
first and second derivative for interpolations (exist sometimes)

**Comment by NAThompson** [comment]
Second derivatives of the quintic B-spline are implemented. Other interpolators have the wrong spectral properties and perform poorly on second derivatives.

2D interpolation is coming! (Akima + pchip)

**Comment by NAThompson** [comment]
@cosurgi : Totally agree it would be useful, but I really like FFTW and it satisfies all my needs, so other things are a higher priority. I don't know who has the skillset to take on such a difficult task . . .

**Comment by mrityunjay-tripathi** [comment]
> What else is a natural fit for the library, and is useful?

* Dirichlet Distribution #317

**Comment by mborland** [comment]
From discussion with the SciPy Maintainers:

- Symplectic ODE solver https://github.com/scipy/scipy/issues/12690
- logcdf and entropy for statistical distributions
- Add support for complex arguments to appropriate statistical distributions

**Comment by mborland** [comment]
> I think the [generalized extreme value distribution](https://en.wikipedia.org/wiki/Generalized_extreme_value_distribution) would be useful addition to the distributions collection.
> 
> I'm working on implementing it. I can submit a PR once I get it working, if there is interest.

We certainly won't say no. Let us know if you have any questions, but I think you're pretty familiar with our internals and testing.

**Comment by WarrenWeckesser** [comment]
Thanks.  I already have a question or two.  Any preference for using the boost developer's mailing list (i.e. https://lists.boost.org/mailman/listinfo.cgi/boost) vs a new github issue?

**Comment by mborland** [comment]
> Thanks. I already have a question or two. Any preference for using the boost developer's mailing list (i.e. https://lists.boost.org/mailman/listinfo.cgi/boost) vs a new github issue?

A GitHub issue would be better.

**Comment by matwey** [comment]
> What else is a natural fit for the library, and is useful?

Single exponential transform quadratures (I know that there are double exponential quadratures in the library, but this research https://link.springer.com/article/10.1007/s00211-008-0195-1 claims that SE are capable for wider class by a price of slower convergence rate. It means that for some integrals where DE works poorly SE can potentially works satisfactorily)
